ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/computation/DistributingExecutorTest.java

```diff
@@ -0,0 +1,119 @@
+import static org.mockito.Mockito.when;
+
+import io.confluent.ksql.parser.KsqlParser.PreparedStatement;
+import io.confluent.ksql.parser.tree.EmptyStatement;
+import io.confluent.ksql.rest.entity.CommandStatus;
+import io.confluent.ksql.rest.entity.CommandStatus.Status;
+import io.confluent.ksql.rest.entity.CommandStatusEntity;
+import io.confluent.ksql.rest.server.computation.CommandId.Action;
+import io.confluent.ksql.rest.server.computation.CommandId.Type;
+import io.confluent.ksql.rest.server.computation.DistributingExecutor.SchemaInference;
+import io.confluent.ksql.services.ServiceContext;
+import io.confluent.ksql.util.KsqlException;
+import java.time.Duration;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class DistributingExecutorTest {
+
+  private static final Duration DURATION_10_MS = Duration.ofMillis(10);
+
+  @Rule
+  public ExpectedException expectedException = ExpectedException.none();
+
+  @Mock CommandQueue queue;
+  @Mock QueuedCommandStatus status;
+  @Mock ServiceContext serviceContext;
+
+  @Test
+  public void shouldEnqueueSuccessfulCommand() throws InterruptedException {
+    // Given:
+    final CommandId command = new CommandId(Type.STREAM, "stream", Action.CREATE);
+    final CommandStatus commandStatus = new CommandStatus(Status.SUCCESS, "");
+
+    when(queue.enqueueCommand(any(), any(), any())).thenReturn(status);
+    when(status.tryWaitForFinalStatus(any())).thenReturn(commandStatus);
+    when(status.getCommandId()).thenReturn(command);
+    when(status.getCommandSequenceNumber()).thenReturn(1L);
+
+    final PreparedStatement preparedStatement = PreparedStatement.of("", new EmptyStatement());
+    final SchemaInference schemaInference = mock(SchemaInference.class);
+    when(schemaInference.apply(any(), any())).thenAnswer(inv -> inv.getArgument(0));
+
+    // When:
+    final CommandStatusEntity commandStatusEntity =
+        (CommandStatusEntity) new DistributingExecutor(queue, DURATION_10_MS, schemaInference)
+            .execute(preparedStatement, null, serviceContext, null, null);
+
+    // Then:
+    assertThat(commandStatusEntity, equalTo(new CommandStatusEntity(
+        "",
+        command,
+        commandStatus,
+        1L
+    )));
+
+    verify(queue, times(1)).enqueueCommand(eq(preparedStatement), any(), any());
+    verify(schemaInference, times(1)).apply(eq(preparedStatement), any());
+  }
+
+  @Test
+  public void testFailingCommandQueue() {
+    // Given:
+    when(queue.enqueueCommand(any(), any(), any())).thenThrow(new KsqlException("Fail!"));
+    final PreparedStatement preparedStatement = PreparedStatement.of("", new EmptyStatement());
+
+    // Expect:
+    expectedException.expect(KsqlException.class);
+    expectedException.expectMessage("Could not write the statement");
```

Comment:
do we lose the actual cause? i.e. "Fail!"

Reply:
This is available in the cause. I have updated the test to reflect this.

Reply:
I *think* the CLI only show the user the first line of a KsqlException. So you might want to include the underlying cause in the outer exception, e.g. "Could not write the statement: fail".

Also, chat with @rodesai to check if this is the correct type of exception to throw for what would be a 'server error' level of failure.  Alternatively, just test what error is returned to the user in this situation.  It should not be a statement error. It should be server error.

Reply:
I introduced `KsqlServerException` to make it clear. In the past, any exception that wasn't of type `KsqlException` would be raised as a server error.